### PR TITLE
fix #16150 improve type mismatch errors

### LINKED ISSUE #16150: Can we improve error message for wrong proc result type?
```nim
type
  O = object
    i: int

proc p(str: string): seq[string] =
  return result

var o: O

o = p("hello")
```

```
nim c --gc:arc t.nim 
....
/tmp/hhh/t.nim(10, 6) Error: type mismatch: got <seq[string]> but expected 'O = object'
salewski@nuc /tmp/hhh $ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2020-11-12
```

In a similar but much more complicated case I got from the error message the impression that the argument of the proc has wrong type. Took me a few minutes to recognize that problem is the proc result.

Error pos column 6 is the start of the proc argument, and when argument and return types are both exotic types it is confusing.

### PR DESCRIPTION
**Pull request by timotheecour**
fix #16150

after PR:

```
Error: type mismatch: got 'seq[string]' for 'p("hello")' but expected 'O = object'
```

with --declaredlocs --listfullpaths:off
```
t11377.nim(14, 8) Error: type mismatch:
 got 'seq[string]' for 'p("hello")' [sequence]
 but expected 'O = object' [object declared in t11377.nim(6, 5)]
```

use -d:nimLegacyTypeMismatch for previous behavior.

## CI failure unrelated: https://github.com/nim-lang/Nim/issues/16169

### CODE REVIEW

File: compiler/semfields.nim [hunk 1]
```diff
@@ -127,9 +127,10 @@ proc semForFields(c: PContext, n: PNode, m: TMagic): PNode =
     localError(c.config, n.info, errGenerated, "no object or tuple type")
     return result
   for i in 1..<call.len:
-    var tupleTypeB = skipTypes(call[i].typ, skippedTypesForFields)
+    let calli = call[i]
+    var tupleTypeB = skipTypes(calli.typ, skippedTypesForFields)
     if not sameType(tupleTypeA, tupleTypeB):
-      typeMismatch(c.config, call[i].info, tupleTypeA, tupleTypeB)
+      typeMismatch(c.config, call[i].info, tupleTypeA, tupleTypeB, calli)
```
Comment (Clyybber) [inline compiler/semfields.nim]: Remove `calli` here and use `call[i]`, or use `calli` consistently.

**Comment by Clyybber** [comment]
I think this is too verbose, consider if/block/case expressions. We should fix the column number instead. (They don't start at 0, try `echo undefined`)

File: compiler/semfields.nim (hunk 1 above)
Comment (timotheecour) [reply compiler/semfields.nim]: done

**Comment by timotheecour** [comment]
@Clyybber 

> I think this is too verbose, consider if/block/case expressions.

behavior as implemented in this PR is consistent with sigmatch errors which also show the expression on error, after template expansion, and regardless of how verbose that can become:
```nim
when true:
  proc fn(a: int) = discard
  proc bar(a: int): auto = 1.0
  var z = 2
  template foo(): untyped = bar(1+z)
  # fn(foo()) # will show an error that contains: but expression 'bar(1 + z)' is of type: float64
  var a: int = foo() # will now show an error containing:  got 'float64' for 'bar(1 + z)' but expected 'int'
```

- [ ] future work can add an option to instead show the expression as it appears in source code (before template expansion), and apply this to all error messages (including type mismatch and sigmatch errors, in particular the rendering of condition in `doAssert` failure messages) instead of showing the expression after template expansion.

This is not trivial to do though (related: issues like https://github.com/nim-lang/Nim/issues/7039 and https://github.com/nim-lang/Nim/issues/11789 for const folding or this: https://github.com/nim-lang/Nim/pull/14863#discussion_r515890665

> Simply remembering both variants can create a compiler that uses an expontential amount of memory. (Nest these calls to get 2x2x2x2x... memory consumption.) There should only be a single AST that works with all the constructs. My "semtityped" AST idea.

There are ways to fix it though.

> We should fix the column number instead. (They don't start at 0, try echo undefined)

- [ ] That should also be fixed, it's an orthogonal issue

**Pull request merged**